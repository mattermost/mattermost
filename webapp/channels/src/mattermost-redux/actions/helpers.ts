// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ActionTypes} from 'mattermost-redux/action_types';
import {GenericAction} from 'mattermost-redux/types/actions';

export function getSharedChannels(): GenericAction {
    return {
        type: ActionTypes.GET_SHARED_CHANNELS,
    };
}

export function receivedSharedChannelsWithRemotes(sharedChannelsWithRemotes: any[]): GenericAction {
    return {
        type: ActionTypes.RECEIVED_SHARED_CHANNELS_WITH_REMOTES,
        data: sharedChannelsWithRemotes,
    };
}