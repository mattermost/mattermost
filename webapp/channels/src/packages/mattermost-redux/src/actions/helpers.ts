// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AnyAction} from 'redux';

import type {ServerError} from '@mattermost/types/errors';
import type {GlobalState} from '@mattermost/types/store';

import {UserTypes} from 'mattermost-redux/action_types';
import {Client4} from 'mattermost-redux/client';
import type {DispatchFunc, GetStateFunc, ActionFuncAsync} from 'mattermost-redux/types/actions';

import {logError} from './errors';

type ActionType = string;
const HTTP_UNAUTHORIZED = 401;
export function forceLogoutIfNecessary(err: ServerError, dispatch: DispatchFunc<AnyAction>, getState: GetStateFunc) {
    const {currentUserId} = getState().entities.users;

    if ('status_code' in err && err.status_code === HTTP_UNAUTHORIZED && err.url && err.url.indexOf('/login') === -1 && currentUserId) {
        Client4.setToken('');
        dispatch({type: UserTypes.LOGOUT_SUCCESS, data: {}});
    }
}

function dispatcher(type: ActionType, data: any, dispatch: DispatchFunc<AnyAction>) {
    if (type.indexOf('SUCCESS') === -1) { // we don't want to pass the data for the request types
        dispatch(requestSuccess(type, data));
    } else {
        dispatch(requestData(type));
    }
}

export function requestData(type: ActionType) {
    return {
        type,
        data: null,
    };
}

export function requestSuccess(type: ActionType, data: any) {
    return {
        type,
        data,
    };
}

export function requestFailure(type: ActionType, error: ServerError): any {
    return {
        type,
        error,
    };
}

/**
 * Returns an ActionFunc which calls a specfied (client) function and
 * dispatches the specifed actions on request, success or failure.
 *
 * @export
 * @param {Object} obj                                       an object for destructirung required properties
 * @param {() => Promise<mixed>} obj.clientFunc              clientFunc to execute
 * @param {ActionType} obj.onRequest                         ActionType to dispatch on request
 * @param {(ActionType | Array<ActionType>)} obj.onSuccess   ActionType to dispatch on success
 * @param {ActionType} obj.onFailure                         ActionType to dispatch on failure
 * @param {...Array<any>} obj.params
 * @returns {ActionFunc} ActionFunc
 */

export function bindClientFunc<Func extends(...args: any[]) => Promise<any>>({
    clientFunc,
    onRequest,
    onSuccess,
    onFailure,
    params,
}: {
    clientFunc: Func;
    onRequest?: ActionType;
    onSuccess?: ActionType | ActionType[];
    onFailure?: ActionType;
    params?: Parameters<Func>;
}): ActionFuncAsync<Awaited<ReturnType<Func>>, GlobalState, AnyAction> {
    return async (dispatch, getState) => {
        if (onRequest) {
            dispatch(requestData(onRequest));
        }

        let data: Awaited<ReturnType<Func>>;
        try {
            if (params) {
                data = await clientFunc(...params);
            } else {
                data = await clientFunc();
            }
        } catch (error) {
            forceLogoutIfNecessary(error, dispatch, getState);
            if (onFailure) {
                dispatch(requestFailure(onFailure, error));
            }
            dispatch(logError(error));
            return {error};
        }

        if (Array.isArray(onSuccess)) {
            onSuccess.forEach((s) => {
                dispatcher(s, data, dispatch);
            });
        } else if (onSuccess) {
            dispatcher(onSuccess, data, dispatch);
        }

        return {data};
    };
}

// Debounce function based on underscores modified to use es6 and a cb

export function debounce(func: (...args: any) => unknown, wait: number, immediate?: boolean, cb?: () => unknown) {
    let timeout: NodeJS.Timeout|null;
    return function fx(...args: any[]) {
        const runLater = () => {
            timeout = null;
            if (!immediate) {
                Reflect.apply(func, null, args);
                if (cb) {
                    cb();
                }
            }
        };
        const callNow = immediate && !timeout;
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(runLater, wait);
        if (callNow) {
            Reflect.apply(func, null, args);
            if (cb) {
                cb();
            }
        }
    };
}

export class FormattedError extends Error {
    intl: {
        id: string;
        defaultMessage: string;
        values: any;
    };

    constructor(id: string, defaultMessage: string, values: any = {}) {
        super(defaultMessage);
        this.intl = {
            id,
            defaultMessage,
            values,
        };
    }
}

