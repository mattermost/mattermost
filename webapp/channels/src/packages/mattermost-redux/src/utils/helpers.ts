// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import shallowEqual from 'shallow-equals';

import {createSelectorCreator, defaultMemoize} from 'mattermost-redux/selectors/create_selector';

// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
export function memoizeResult<F extends Function>(func: F, measure: Function | undefined = undefined): F {
    let lastArgs: IArguments|null = null;
    let lastResult: any = null;

    // we reference arguments instead of spreading them for performance reasons
    return function memoizedFunc() {
        if (!shallowEqual(lastArgs, arguments)) { //eslint-disable-line prefer-rest-params
            //eslint-disable-line prefer-rest-params
            // apply arguments instead of spreading for performance.
            const result = Reflect.apply(func, null, arguments); //eslint-disable-line prefer-rest-params
            if (!shallowEqual(lastResult, result)) {
                lastResult = result;
            }
        }

        if (measure) {
            measure();
        }

        lastArgs = arguments; //eslint-disable-line prefer-rest-params
        return lastResult;
    } as unknown as F;
}

// Use this selector when you want a shallow comparison of the arguments and you want to memoize the result
// try and use this only when your selector returns an array of ids
export const createIdsSelector = createSelectorCreator(memoizeResult);

// Use this selector when you want a shallow comparison of the arguments and you don't need to memoize the result
export const createShallowSelector = createSelectorCreator(defaultMemoize, shallowEqual as any);

// isMinimumServerVersion will return true if currentVersion is equal to higher or than the
// the provided minimum version. A non-equal major version will ignore minor and dot
// versions, and a non-equal minor version will ignore dot version.
// currentVersion is a string, e.g '4.6.0'
// minMajorVersion, minMinorVersion, minDotVersion are integers
export const isMinimumServerVersion = (currentVersion: string, minMajorVersion = 0, minMinorVersion = 0, minDotVersion = 0): boolean => {
    if (!currentVersion || typeof currentVersion !== 'string') {
        return false;
    }

    const split = currentVersion.split('.');

    const major = parseInt(split[0], 10);
    const minor = parseInt(split[1] || '0', 10);
    const dot = parseInt(split[2] || '0', 10);

    if (major > minMajorVersion) {
        return true;
    }
    if (major < minMajorVersion) {
        return false;
    }

    // Major version is equal, check minor
    if (minor > minMinorVersion) {
        return true;
    }
    if (minor < minMinorVersion) {
        return false;
    }

    // Minor version is equal, check dot
    if (dot > minDotVersion) {
        return true;
    }
    if (dot < minDotVersion) {
        return false;
    }

    // Dot version is equal
    return true;
};

// Generates a RFC-4122 version 4 compliant globally unique identifier.
export function generateId(): string {
    // implementation taken from http://stackoverflow.com/a/2117523
    let id = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
    id = id.replace(/[xy]/g, (c) => {
        const r = Math.floor(Math.random() * 16);
        let v;

        if (c === 'x') {
            v = r;
        } else {
            // eslint-disable-next-line no-mixed-operators
            v = r & 0x3 | 0x8;
        }

        return v.toString(16);
    });
    return id;
}

export function isEmail(email: string): boolean {
    // writing a regex to match all valid email addresses is really, really hard. (see http://stackoverflow.com/a/201378)
    // this regex ensures:
    // - at least one character that is not a space, comma, or @ symbol
    // - followed by a single @ symbol
    // - followed by at least one character that is not a space, comma, or @ symbol
    // this prevents <Outlook Style> outlook.style@domain.com addresses and multiple comma-separated addresses from being accepted
    return (/^[^ ,@]+@[^ ,@]+$/).test(email);
}
