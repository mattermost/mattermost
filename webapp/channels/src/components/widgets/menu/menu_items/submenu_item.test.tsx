// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {mountWithIntl} from 'tests/helpers/intl-test-helper';
import {screen, userEvent, renderWithContext} from 'tests/react_testing_utils';
import Constants from 'utils/constants';

import SubMenuItem, {SubMenuItem as SubMenuItemClass} from './submenu_item';

jest.mock('../is_mobile_view_hack', () => ({
    isMobile: jest.fn(() => false),
}));

describe('components/widgets/menu/menu_items/submenu_item', () => {
    test('empty subMenu should match snapshot', () => {
        const wrapper = mountWithIntl(
            <SubMenuItem
                key={'_pluginmenuitem'}
                id={'1'}
                text={'test'}
                subMenu={[]}
                action={jest.fn()}
                root={true}
            />,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('present subMenu should match snapshot with submenu', () => {
        const wrapper = mountWithIntl(
            <SubMenuItem
                key={'_pluginmenuitem'}
                id={'1'}
                text={'test'}
                subMenu={[
                    {
                        id: 'A',
                        text: 'Test A',
                        direction: 'left',
                    },
                    {
                        id: 'B',
                        text: 'Test B',
                        direction: 'left',
                    },
                ]}
                action={jest.fn()}
                root={true}
            />,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('test subMenu click triggers action', () => {
        const action1 = jest.fn();
        const action2 = jest.fn();
        const action3 = jest.fn();

        renderWithContext(
            <SubMenuItem
                key={'_pluginmenuitem'}
                id={'Z'}
                text={'test'}
                subMenu={[
                    {
                        id: 'A',
                        text: 'Test A',
                        action: action2,
                        direction: 'left',
                    },
                    {
                        id: 'B',
                        text: 'Test B',
                        action: action3,
                        direction: 'left',
                    },
                ]}
                action={action1}
                root={true}
            />,
        );

        userEvent.click(screen.getByText('test'));
        expect(action1).toHaveBeenCalledTimes(1);

        userEvent.click(screen.getByText('Test A'));
        expect(action2).toHaveBeenCalledTimes(1);

        userEvent.click(screen.getByText('Test B'));
        expect(action3).toHaveBeenCalledTimes(1);

        // Confirm that the parent's action wasn't called again when clicking on a child item
        expect(action1).toHaveBeenCalledTimes(1);
    });

    test('should show/hide submenu based on keyboard commands', () => {
        const wrapper = mountWithIntl(
            <SubMenuItem
                key={'_pluginmenuitem'}
                id={'1'}
                text={'test'}
                subMenu={[]}
                root={true}
                direction={'right'}
            />,
        );

        const instance = wrapper.find(SubMenuItemClass).instance() as SubMenuItemClass;

        instance.show = jest.fn();
        instance.hide = jest.fn();

        instance.handleKeyDown({keyCode: Constants.KeyCodes.ENTER[1]} as any);
        expect(instance.show).toHaveBeenCalled();

        instance.handleKeyDown({keyCode: Constants.KeyCodes.LEFT[1]} as any);
        expect(instance.hide).toHaveBeenCalled();

        instance.handleKeyDown({keyCode: Constants.KeyCodes.RIGHT[1]} as any);
        expect(instance.show).toHaveBeenCalled();
    });
});
