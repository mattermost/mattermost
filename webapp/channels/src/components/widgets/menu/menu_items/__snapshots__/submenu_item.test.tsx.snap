// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/widgets/menu/menu_items/submenu_item empty subMenu should match snapshot 1`] = `
<SubMenuItem
  action={[MockFunction]}
  direction="left"
  id="1"
  intl={
    Object {
      "$t": [Function],
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "defaultRichTextElements": undefined,
      "fallbackOnEmptyString": true,
      "formatDate": [Function],
      "formatDateTimeRange": [Function],
      "formatDateToParts": [Function],
      "formatDisplayName": [Function],
      "formatList": [Function],
      "formatListToParts": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatNumberToParts": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "formatTimeToParts": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getDisplayNames": [Function],
        "getListFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralRules": [Function],
        "getRelativeTimeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "onError": [Function],
      "onWarn": [Function],
      "textComponent": "span",
      "timeZone": "Etc/UTC",
    }
  }
  renderSelected={true}
  root={true}
  show={true}
  subMenu={Array []}
  subMenuClass="pl-4"
  text="test"
>
  <li
    className="SubMenuItem MenuItem"
    id="1_menuitem"
    onClick={[Function]}
    role="menuitem"
  >
    <div
      className=""
      id="1"
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      tabIndex={0}
    >
      <div
        className="flex"
      >
        test
        <span
          className="selected"
        />
        <span
          aria-label="submenu icon"
          className="fa fa-angle-right SubMenu__icon-right-empty"
          id="channelHeaderDropdownIconRight_1"
        />
      </div>
      <ul
        className="a11y__popup Menu dropdown-menu SubMenu"
        style={
          Object {
            "right": "100%",
            "top": "unset",
            "visibility": "hidden",
          }
        }
      />
    </div>
  </li>
</SubMenuItem>
`;

exports[`components/widgets/menu/menu_items/submenu_item present subMenu should match snapshot with submenu 1`] = `
<SubMenuItem
  action={[MockFunction]}
  direction="left"
  id="1"
  intl={
    Object {
      "$t": [Function],
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "defaultRichTextElements": undefined,
      "fallbackOnEmptyString": true,
      "formatDate": [Function],
      "formatDateTimeRange": [Function],
      "formatDateToParts": [Function],
      "formatDisplayName": [Function],
      "formatList": [Function],
      "formatListToParts": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatNumberToParts": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "formatTimeToParts": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getDisplayNames": [Function],
        "getListFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralRules": [Function],
        "getRelativeTimeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "onError": [Function],
      "onWarn": [Function],
      "textComponent": "span",
      "timeZone": "Etc/UTC",
    }
  }
  renderSelected={true}
  root={true}
  show={true}
  subMenu={
    Array [
      Object {
        "direction": "left",
        "id": "A",
        "text": "Test A",
      },
      Object {
        "direction": "left",
        "id": "B",
        "text": "Test B",
      },
    ]
  }
  subMenuClass="pl-4"
  text="test"
>
  <li
    className="SubMenuItem MenuItem"
    id="1_menuitem"
    onClick={[Function]}
    role="menuitem"
  >
    <div
      className=""
      id="1"
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      tabIndex={0}
    >
      <div
        className="flex"
      >
        test
        <span
          className="selected"
        />
        <span
          aria-label="submenu icon"
          className="fa fa-angle-right SubMenu__icon-right"
          id="channelHeaderDropdownIconRight_1"
        />
      </div>
      <ul
        className="a11y__popup Menu dropdown-menu SubMenu"
        style={
          Object {
            "right": "100%",
            "top": "unset",
            "visibility": "hidden",
          }
        }
      >
        <span
          className="SubMenuItemContainer"
          key="A"
          tabIndex={0}
        >
          <SubMenuItem
            direction="left"
            id="A"
            intl={
              Object {
                "$t": [Function],
                "defaultFormats": Object {},
                "defaultLocale": "en",
                "defaultRichTextElements": undefined,
                "fallbackOnEmptyString": true,
                "formatDate": [Function],
                "formatDateTimeRange": [Function],
                "formatDateToParts": [Function],
                "formatDisplayName": [Function],
                "formatList": [Function],
                "formatListToParts": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatNumberToParts": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "formatTimeToParts": [Function],
                "formats": Object {},
                "formatters": Object {
                  "getDateTimeFormat": [Function],
                  "getDisplayNames": [Function],
                  "getListFormat": [Function],
                  "getMessageFormat": [Function],
                  "getNumberFormat": [Function],
                  "getPluralRules": [Function],
                  "getRelativeTimeFormat": [Function],
                },
                "locale": "en",
                "messages": Object {},
                "onError": [Function],
                "onWarn": [Function],
                "textComponent": "span",
                "timeZone": "Etc/UTC",
              }
            }
            renderSelected={true}
            root={false}
            show={true}
            subMenuClass="pl-4"
            tabIndex={1}
            text="Test A"
          >
            <li
              className="SubMenuItem MenuItem"
              id="A_menuitem"
              onClick={[Function]}
              role="menuitem"
            >
              <div
                className=""
                id="A"
                onKeyDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                tabIndex={1}
              >
                <div
                  className="flex"
                >
                  Test A
                  <span
                    className="selected"
                  />
                  <span
                    aria-label="submenu icon"
                    className="fa fa-angle-right SubMenu__icon-right-empty"
                    id="channelHeaderDropdownIconRight_A"
                  />
                </div>
                <ul
                  className="a11y__popup Menu dropdown-menu SubMenu"
                  style={
                    Object {
                      "right": "100%",
                      "top": "unset",
                      "visibility": "hidden",
                    }
                  }
                />
              </div>
            </li>
          </SubMenuItem>
        </span>
        <span
          className="SubMenuItemContainer"
          key="B"
          tabIndex={0}
        >
          <SubMenuItem
            direction="left"
            id="B"
            intl={
              Object {
                "$t": [Function],
                "defaultFormats": Object {},
                "defaultLocale": "en",
                "defaultRichTextElements": undefined,
                "fallbackOnEmptyString": true,
                "formatDate": [Function],
                "formatDateTimeRange": [Function],
                "formatDateToParts": [Function],
                "formatDisplayName": [Function],
                "formatList": [Function],
                "formatListToParts": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatNumberToParts": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "formatTimeToParts": [Function],
                "formats": Object {},
                "formatters": Object {
                  "getDateTimeFormat": [Function],
                  "getDisplayNames": [Function],
                  "getListFormat": [Function],
                  "getMessageFormat": [Function],
                  "getNumberFormat": [Function],
                  "getPluralRules": [Function],
                  "getRelativeTimeFormat": [Function],
                },
                "locale": "en",
                "messages": Object {},
                "onError": [Function],
                "onWarn": [Function],
                "textComponent": "span",
                "timeZone": "Etc/UTC",
              }
            }
            renderSelected={true}
            root={false}
            show={true}
            subMenuClass="pl-4"
            tabIndex={1}
            text="Test B"
          >
            <li
              className="SubMenuItem MenuItem"
              id="B_menuitem"
              onClick={[Function]}
              role="menuitem"
            >
              <div
                className=""
                id="B"
                onKeyDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                tabIndex={1}
              >
                <div
                  className="flex"
                >
                  Test B
                  <span
                    className="selected"
                  />
                  <span
                    aria-label="submenu icon"
                    className="fa fa-angle-right SubMenu__icon-right-empty"
                    id="channelHeaderDropdownIconRight_B"
                  />
                </div>
                <ul
                  className="a11y__popup Menu dropdown-menu SubMenu"
                  style={
                    Object {
                      "right": "100%",
                      "top": "unset",
                      "visibility": "hidden",
                    }
                  }
                />
              </div>
            </li>
          </SubMenuItem>
        </span>
      </ul>
    </div>
  </li>
</SubMenuItem>
`;
